packages/ai/core/tool/mcp/client.ts

```diff
@@ -0,0 +1,318 @@
+          params: {
+            protocolVersion: LATEST_PROTOCOL_VERSION,
+            capabilities: {},
+            clientInfo: this._clientInfo,
+          },
+        },
+        InitializeResultSchema,
+        { timeout: this.connectionTimeoutMs },
+      );
+
+      if (result === undefined) {
+        throw new AISDKError({
+          name: 'McpClientError',
+          message: 'Server sent invalid initialize result',
+        });
+      }
+
+      if (!SUPPORTED_PROTOCOL_VERSIONS.includes(result.protocolVersion)) {
+        throw new AISDKError({
+          name: 'McpClientError',
+          message: `Server's protocol version is not supported: ${result.protocolVersion}`,
+        });
+      }
+
+      this._serverCapabilities = result.capabilities;
+
+      await this.notification({
+        method: 'notifications/initialized',
+      });
+    } catch (error) {
+      void this.close();
+      throw error;
+    }
+  }
+
+  async close(): Promise<void> {
+    await this._transport?.close();
+    this._onclose();
+  }
+
+  async request<T extends ZodType<object>>(
+    request: Request,
+    resultSchema: T,
+    options?: RequestOptions,
+  ): Promise<z.infer<T>> {
+    return new Promise((resolve, reject) => {
+      if (!this._transport) {
+        reject(new Error('Not connected'));
+        return;
+      }
+
+      options?.signal?.throwIfAborted();
+
+      const messageId = this._requestMessageId++;
+      const jsonrpcRequest: JSONRPCRequest = {
+        ...request,
+        jsonrpc: '2.0',
+        id: messageId,
+      };
+
+      const cancel = (reason: unknown) => {};
+
+      this._responseHandlers.set(messageId, response => {
+        if (options?.signal?.aborted) return;
+
+        if (response instanceof Error) {
+          return reject(response);
+        }
+
+        try {
+          const result = resultSchema.parse(response.result);
+          resolve(result);
+        } catch (error) {
+          reject(error);
+        }
+      });
+
+      options?.signal?.addEventListener('abort', () => {
+        cancel(options?.signal?.reason);
+      });
+
+      const timeout = options?.timeout ?? this.requestTimeoutMs;
+      const timeoutHandler = () =>
+        cancel(
+          new AISDKError({
+            name: 'McpClientTimeoutError',
+            message: 'Request timed out',
+          }),
+        );
+
+      this._setupTimeout(
+        messageId,
+        timeout,
+        options?.maxTotalTimeout,
+        timeoutHandler,
+      );
+
+      this._transport.send(jsonrpcRequest).catch(error => {
+        this._cleanupTimeout(messageId);
+        reject(error);
+      });
+    });
+  }
+
+  async listTools(
+    params?: ListToolsRequest['params'],
+    options?: RequestOptions,
+  ): Promise<ListToolsResult> {
+    if (!this._serverCapabilities?.tools) {
+      throw new AISDKError({
+        name: 'McpClientError',
+        message: `Server does not support tools (required for tools/list)`,
+      });
+    }
```

Comment:
we should not have that flag. always assume tools are supported (or the server is useless for the ai sdk)